### PR TITLE
Added functionality for material pipeline files to specify custom object SRG members

### PR DESCRIPTION
**Pull request by antonmic**
Now in your .materialpipeline file you can write:

"objectSrg": [
    "float4 m_myCustomVar1",
    "uint   m_myCustomVar2"
],

and the material type builder will add m_myCustomVar1 and m_myCustomVar2
to the ObjectSrg of all materials rendered in your material pipeline

This allows custom gems and material pipelines to add necessary members to the object SRG without touching core .azsli files

### CODE REVIEW

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialTypeBuilder.cpp [hunk 1]
```diff
@@ -43,7 +43,7 @@ namespace AZ
         {
             AssetBuilderSDK::AssetBuilderDesc materialBuilderDescriptor;
             materialBuilderDescriptor.m_name = "Material Type Builder";
-            materialBuilderDescriptor.m_version = 38; // material type indirect references
+            materialBuilderDescriptor.m_version = 42; // shader defines from material pipelines
```
Comment (galibzon) [inline Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialTypeBuilder.cpp:46]: This should be 39.

File: Gems/Atom/RPI/Code/Source/RPI.Edit/Material/MaterialPipelineSourceData.cpp [hunk 2]
```diff
@@ -34,10 +34,11 @@ namespace AZ
                     ;
 
                 serializeContext->Class<MaterialPipelineSourceData>()
-                    ->Version(1)
+                    ->Version(4)    // Object Srg Additions
```
Comment (galibzon) [inline Gems/Atom/RPI/Code/Source/RPI.Edit/Material/MaterialPipelineSourceData.cpp:37]: Version 2?

**Review by galibzon — APPROVED** [review]

**Review by gadams3 — COMMENTED** [review]
Are there any other built-in options for extending an SRG?
If not, does the compiler support adding include statements inside of the SRG?
It would be much more flexible and easier to edit an optional include file for the SRG extensions.
It would allow using a text editor, not dealing with escape characters, as well as adding comments and functions if needed.
The same is true for material canvas, where it is easy to embed some simple lines directly in node config files but more complex functions should be in dedicated include files for the same reasons mentioned above.
The AZSLI files would pretty much look like what you have already but MATERIAL_PIPELINE_OBJECT_SRG_MEMBERS would resolve to an include statement.
You would not need to modify the material pipeline structure.
You could use a separate text file in the same folder as the material pipeline file, possibly with the same name as the material pipeline file plus a suffix _ObjectSRG-EXT.azsli.
Then the material builder checks for the files existence and sets the define to include that absolute path.

**Review by gadams3 — APPROVED** [review]
Discussed offline. Please follow up with comments that the strings are intended for simple variables and will need character escaping for new lines, possibly comments, and other characters if they are embedded in the strings.

**Pull request merged**